libraries/AC_SafeRTL/AC_SafeRTL.cpp

```diff
@@ -0,0 +1,573 @@
+    reset_simplification(true);
+    // reset pruning
+    reset_pruning(true);
+
+    if (!position_ok) {
+        _active = false;
+        log_action(SRTL_DEACTIVATED_BAD_POSITION);
+        gcs().send_text(MAV_SEVERITY_WARNING, "SafeRTL: unavailable bad position");
+        return;
+    }
+
+    // constrain the path length, in case the user decided to make the path unreasonably long.
+    _desired_path_len = MIN(SAFERTL_PATH_LEN_MAX, _desired_path_len);
+
+    _last_index = 0;
+    _path[_last_index] = current_pos;
+    _time_of_last_good_position = AP_HAL::millis();
+    _active = true;
+}
+
+// call this a couple of times per second regardless of what mode the vehicle is in
+void AC_SafeRTL::update(bool position_ok, bool save_position)
+{
+    if (!_active || !position_ok || !save_position) {
+        return;
+    }
+
+    Vector3f current_pos;
+    position_ok &= _ahrs.get_relative_position_NED_origin(current_pos);
+    update(position_ok, current_pos);
+}
+
+void AC_SafeRTL::update(bool position_ok, const Vector3f& current_pos)
+{
+    if (!_active) {
+        return;
+    }
+
+    if (position_ok) {
+        _time_of_last_good_position = AP_HAL::millis();
+    } else {
+        if (AP_HAL::millis() - _time_of_last_good_position > SAFERTL_BAD_POSITION_TIMEOUT) {
+            _active = false;
+            log_action(SRTL_DEACTIVATED_BAD_POSITION);
+            gcs().send_text(MAV_SEVERITY_WARNING,"SafeRTL: unavailable bad position");
+        }
+        return;
+    }
+
+    // it's important to do the cleanup before adding the point, because appending a point will reset the cleanup methods,
+    // so there will not be anything to clean up immediately after adding a point.
+    // The cleanup usually returns immediately. If it decides to actually perform the cleanup, it takes about 100us.
+    if (!routine_cleanup()) {
+        _active = false;
+        log_action(SRTL_DEACTIVATED_CLEANUP_FAILED);
+        gcs().send_text(MAV_SEVERITY_WARNING,"SafeRTL: unavailable path cleanup failed");
+        return;
+    }
+
+    if (position_ok) {
+        // append the new point, if appropriate
+        if (HYPOT(current_pos, _path[_last_index]) > _accuracy) {
+            // add the breadcrumb
+            _path[++_last_index] = current_pos;
+
+            log_action(SRTL_POINT_ADD, current_pos);
+
+            // if cleanup algorithms are finished (and therefore not running), reset them
+            if (_simplification_complete) {
+                reset_simplification(false);
+            }
+            if (_pruning_complete) {
+                reset_pruning(false);
+            }
+        }
+    }
+}
+
+/**
+*  Run this method only when preparing to initiate the RTL procedure. Returns a
+*  pointer to the cleaned-up path. Returns nullptr if the cleanup algorithms aren't ready yet.
+*  If this happens, just run this method again a bit later.
+*
+*  Probably best not to run this unless cleanup_ready() is returning true
```

Comment:
I think this is an out of date comment.  It has the checks which confirm the background cleanup tasks have completed.